### PR TITLE
Try longer to fetch initial token.

### PR DESCRIPTION
**Pull request by porridge**
**What this PR does / why we need it**:
Step towards fixing #56293

**Special notes for your reviewer**:
/kind bug
/priority critial-urgent
@kubernetes/sig-scalability-bugs
/cc @shyamjvs please add to v1.9

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by shyamjvs** [comment]
LGTM - but do we know if the reason for that failure is a transient error or sth else?

**Comment by porridge** [comment]
Unfortunately we don't.  I suspect it's a local network misconfiguration,
because I don't think the service would be down.

File: pkg/cloudprovider/providers/gce/gce.go [hunk 1]
```diff
@@ -810,7 +810,7 @@ func newOauthClient(tokenSource oauth2.TokenSource) (*http.Client, error) {
 		glog.Infof("Using existing Token Source %#v", tokenSource)
 	}
 
-	if err := wait.PollImmediate(5*time.Second, 30*time.Second, func() (bool, error) {
+	if err := wait.PollImmediate(5*time.Second, 5*time.Minute, func() (bool, error) {
```
Comment (mikedanese) [inline pkg/cloudprovider/providers/gce/gce.go]: Have we asked anyone why we should need this many retries? It seems like this should very rarely fail on the first try.

**Comment by mikedanese** [comment]
The error appears to be happening from our CI clusters using gce client credentials. Are our test projects getting throttled?

**Comment by porridge** [comment]
@mikedanese as you can see in the linked issue, the error is during TLS handshake with accounts.google.com - well before we provide the service with anything that has to do with our project.

**Comment by porridge** [comment]
ping @mikedanese 
/test pull-kubernetes-e2e-gce

**Comment by spiffxp** [comment]
needs rebase (the verify job is failing on 2017 vs. 2018)

**Comment by porridge** [comment]
rebased
ping reviewers @mikedanese @rootfs

**Comment by fejta** [comment]
TLS handshake issues have historically been associated with CPU starvation

File: pkg/cloudprovider/providers/gce/gce.go (hunk 1 above)
Comment (ixdy) [reply pkg/cloudprovider/providers/gce/gce.go]: yeah, I'm also a bit skeptical that something that fails after 30s of retries will pass with 5m of retries.

**Comment by porridge** [comment]
Well, my workstation (which is where the client ran) certainly wasn't CPU-starved at that point, and certainly not for 30 seconds.
The point of this change is to retry long enough that if it fails again, we have a good case for escalating to the owners of whatever endpoint this is trying to connect to.

**Comment by porridge** [comment]
ping reviewers @mikedanese @rootfs again

**Comment by mikedanese** [comment]
If we can't get a token for 30 seconds something is terribly wrong. I am not happy with blindly multiplying the timeout by 10.

**Comment by porridge** [comment]
@mikedanese please take a look at the associated bug, frankly I don't see any other option: this happened twice so far when setting up a test cluster from continuous integration. The only reasons that I can think of are:
1. a transient outage of accounts.google.com
2. transient network issues
3. severe CPU starvation on the client side (very unlikely in this case IMO)

In all of these cases, what options do we have other than retrying longer? If we fail after 30 seconds, then knowing that there was a problem during a 30 second window is not nearly enough to investigate the root cause, service owners will likely just shrug.

OTOH being able to prove that a TLS connection could not be established for 5 minutes probably needs to be investigated with care, and it is more likely that we will for example have CPU usage stats with a 5-minute resolution after a while.

**Comment by porridge** [comment]
OK, I did some digging in old logs and this is more frequent than I thought. So the "terribly wrong" may just be the norm in fact.

Just looking at a single sig-scalability CI job, it happened [11 times in the last 3 months or so](https://github.com/kubernetes/kubernetes/issues/56293#issuecomment-361565918). And when our 4 retries are not enough (~30% of cases so far), we just fail the test, adding to the [long list of transient failures](https://github.com/kubernetes/kubernetes/issues/56062) that I'm trying to cut down on.

@mikedanese I don't mind using a threshold slightly lower than 5 minutes, but 30 seconds is clearly too little.

**Comment by shyamjvs** [comment]
I agree with @porridge here. This seems to be hurting more than occasionally.

**Comment by mikedanese** [comment]
Can we do one minute with jittered exponential backoff? The wait libraries should support that.

**Comment by porridge** [comment]
@mikedanese done, I hope I understood you correctly.

File: pkg/cloudprovider/providers/gce/gce.go [hunk 2]
```diff
@@ -827,7 +827,13 @@ func newOauthClient(tokenSource oauth2.TokenSource) (*http.Client, error) {
 		glog.Infof("Using existing Token Source %#v", tokenSource)
 	}
 
-	if err := wait.PollImmediate(5*time.Second, 30*time.Second, func() (bool, error) {
+	backoff := wait.Backoff{
+		// These values will add up to about a minute. See #56293 for background.
+		Duration: time.Second,
+		Factor:   1.4,
+		Steps:    10,
```
Comment (mikedanese) [inline pkg/cloudprovider/providers/gce/gce.go]: Can you also add a jitter parameter? maybe 0.2

**Comment by mikedanese** [comment]
Please add jitter. Otherwise

/lgtm
/approve
/hold

**Comment by porridge** [comment]
A default jitter is used if unspecified in the Backoff struct.
/hold cancel

**Comment by mikedanese** [comment]
Thanks.

**Pull request merged**